hoomd/md/NeighborList.cc

```diff
@@ -366,6 +366,19 @@ void NeighborList::compute(uint64_t timestep)
     if (m_prof)
         m_prof->push("Neighbor");
 
+    // when the number of particles in the system changes, rebuild the exclusion list
+    if (m_need_reallocate_exlist)
+        {
+        clearExclusions();
+        std::set<std::string> exclusions_copy(m_exclusions);
+        for (const std::string& exclusion : exclusions_copy)
+            {
+            setSingleExclusion(exclusion);
+            }
+
+        m_need_reallocate_exlist = false;
```

Comment:
I think this is probably just because I'm not very familiar with this part of the code, but I find something vaguely unsettling about this block right now. Perhaps you could clarify a few things for me:

It looks like this flag is being set by the particle number change signal handler. That seems sensible, but it looks like `clearExclusions` also uses that check for _part_ of its logic and unsets `m_need_reallocate_exlist` at the end (my understanding is that it clears the list unconditionally but only reallocates if this flag is set, indicating that e.g. the size of the exclusion list has to change to accommodate a change in the number of particles). It makes sense to condition this block on `if (m_need_reallocate_exlist)` since in this case you don't want to clear the exclusions unconditionally, but that raises the following questions for me:
1. Is there any situation where the code could hit `clearExclusions` after the particle number changes, but before `compute` is called, such that the exclusion list is cleared but `setSingleExclusion` is not called in this loop?
2. The two differences I see between this look and the `setExclusions` method are that `setFilterBody(false)` call and the fact that this flag is not being set in that method. Is the former difference meaningful? Should `setExclusions` be checking that the argument list length matches the current exclusion list length, or is that method purely an internal-facing method for modifying without resizing the exclusion list?
3. Could `setExclusions` be modified to accept an iterator rather than a `pybind::list` and then reused here? 
4. Setting this variable to false here is superfluous since `clearExclusions` does so as well. I assume that was just a small missed redundancy, but just want to double-check in light of my other questions: is there any reason we wouldn't want to rely on that behavior of `clearExclusions`?

Sorry to drag out review of a PR that generally looks fine and is very small, I just want to make sure I understand what's happening here.

Reply:
@vyasr The C++ code was originally written to interface with the v2 python API. In v2, in the setup phase of every call to `run`, Python would call `clearExclusions` and then call the relevant `addExclusionsFrom*` methods. Thus, the number of particles was only allowed to change in between calls to `run`.

In v3, the number of particles can change during a run as well, and there is no setup phase to the `run` method. When we converted the neighborlist over to the v3 API, we failed to handle updates to the exclusion list. This failure resulted in a debug assertion which was reported in #1029.

1. No, there is no case where `clearExclusions` would be called first. We could rename the function to `resizeAndClearExclusions`, make it private, and rename `m_need_reallocate_exlist` to make the meaning of the variable more clear. The intention is to delay the reallocation  (and subsequent reinitialization) of the data structure until the last possible moment just before it is needed.

2. When the list is reset due to a size change, the body filtering boolean is not changed. If the user changes that flag, it will trigger a new call to `setExclusions` which will update the flag accordingly.

3. I don't know.

4. See (1).

Reply:
Got it thanks for the explanations. I think this is good to go then.